### PR TITLE
Introduce AT_FORALL_QINTS

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23340 [WIP] Remove unused parameter from FORALL macros and rename STUBS to QINTS.
* #23339 [WIP] Align AT_FORALL macros with AT_DISPATCH macros.
* #23338 [WIP] Don't redefine unecessary type stub.
* #23337 [WIP] Support torch::tensor and at::tensor with bool and BFloat16 dtypes.
* #23336 [WIP] Rename AT_FORALL_SCALAR_TYPES_WITH_COMPLEX to AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_STUBS
* #22932 Remove AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF, which isn't used anymore.
* **#22931 Dont allow conversions with QInt.**
* #22890 Remove cases of AT_FORALL_SCALAR_TYPES_EXCEPT_HALF.
* #22889 C++ at::Tensor, torch::tensor constructors should not accept QInts.

Differential Revision: [D16467985](https://our.internmc.facebook.com/intern/diff/D16467985)

### CODE REVIEW

File: aten/src/ATen/core/TensorMethods.h [hunk 1]
```diff
@@ -1800,7 +1800,8 @@ inline bool is_quantized(Tensor self) {
     return static_cast<T*>(this->data_ptr());    \
   }
 
-AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF(DEFINE_CAST)
+AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF_AND_QINT(DEFINE_CAST)
+AT_FORALL_QINTS(DEFINE_CAST)
```
Comment (jerryzh168) [inline aten/src/ATen/core/TensorMethods.h]: do we want to cast them to quint8* or uint8*?

Comment (jerryzh168) [reply aten/src/ATen/core/TensorMethods.h]: I feel `uint8*` might be more useful for people

Comment (gchanan) [reply aten/src/ATen/core/TensorMethods.h]: ya, I think that's the correct solution but outside of the scope of this PR, which is just trying to rationalize the AT_FORALL macros.  So I'm just trying to match up with this one: https://github.com/pytorch/pytorch/blob/34f53564b4e5937780a9784e7694c4df98c217cb/aten/src/ATen/Dispatch.h#L256-L269

so I should have actually called this AT_FORALL_QINT_TYPES, but I fix that up in a later PR in this stack.

Comment (gchanan) [reply aten/src/ATen/core/TensorMethods.h]: I filed https://github.com/pytorch/pytorch/issues/23440 for addressing that.

**Review by jerryzh168 — APPROVED** [review]
LGTM

**Pull request closed without merging**